### PR TITLE
fix: preserve expression names when replacing placeholders

### LINKED ISSUE #12065: Aggregations with parametrized FILTER WHERE clauses don't work
### Describe the bug

Queries with FILTER WHERE don't work when parameters are provided.

For instance, `SELECT SUM (amount) FILTER (WHERE year = $1) FROM payments` will fail on physical planning.

Query like `COUNT (CASE WHEN year = $1 then amount ELSE NULL END)` fails as well.

### To Reproduce

```rust
async fn main() -> Result<()> {
    //  generate test memtable with 30 u64 columns and some data
    let table1 = generate_memtable(30)?;

    let cfg = SessionConfig::new()
        .set("datafusion.sql_parser.dialect", "PostgreSQL".into());

    let ctx = SessionContext::new_with_config(cfg);

    ctx.register_table("table1", Arc::new(table1))?;

    let dataframe = ctx.sql(
        "select count (f_0) filter (where f_1 > $1) from table1"
        ).await?
        .with_param_values(ParamValues::List(vec![ScalarValue::from(10u64)]));

    // fails here on physical planning:
    let result = dataframe?.collect().await.unwrap();

    let record_batch = result.first().unwrap();

    dbg!(record_batch.columns());

    Ok(())
}
```

This fails with error:

```
thread 'main' panicked at filter-where-repro.rs:46:45:
called `Result::unwrap()` on an `Err` value: 
Context("type_coercion", SchemaError(FieldNotFound { field: Column { relation: None, name: "count(table1.f_0) FILTER (WHERE table1.f_1 > $1)" }, valid_fields: [Column { relation: None, name: "count(table1.f_0) FILTER (WHERE table1.f_1 > UInt64(10))" }] }, Some("")))
```

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Closes #12065.

## Rationale for this change
When replacing placeholders with literal values, the names of the expressions might change, but there are references to the old names in the parent plan.

## What changes are included in this PR?
- Preserve expression names when replacing placeholders.
- Move `NamePreserver` to `datafusion-expr` package because it is needed there.

## Are these changes tested?
Yes

## Are there any user-facing changes?

After replacing the parameters, the logical plan might differ from before because aliases have been added.

### CODE REVIEW

File: datafusion/core/src/dataframe/mod.rs [hunk 1]
```diff
@@ -3699,4 +3699,32 @@ mod tests {
         assert!(result.is_err());
         Ok(())
     }
+
+    // Test issue: https://github.com/apache/datafusion/issues/12065
+    #[tokio::test]
+    async fn filtered_aggr_with_param_values() -> Result<()> {
```
Comment (jonahgao) [inline datafusion/core/src/dataframe/mod.rs:3705]: This test on the main branch gave the following error
```
Error: Context("type_coercion", SchemaError(FieldNotFound { field: 
    Column { relation: None, name: "count(table1.c2) FILTER (WHERE table1.c3 > $1)" }, 

valid_fields: [
    Column { relation: None, name: "count(table1.c2) FILTER (WHERE table1.c3 > UInt64(10))" }
] }, Some("")))
```

File: datafusion/optimizer/src/utils.rs [hunk 2]
```diff
@@ -28,6 +28,10 @@ use datafusion_expr::{logical_plan::LogicalPlan, Expr, Operator};
 
 use log::{debug, trace};
 
+/// Re-export of `NamesPreserver` for backwards compatibility,
+/// as it was initially placed here and then moved elsewhere.
```
Comment (alamb) [inline datafusion/optimizer/src/utils.rs:32]: 👍

File: datafusion/expr/src/expr_rewriter/mod.rs [hunk 3]
```diff
@@ -279,6 +279,57 @@ where
     expr.alias_if_changed(original_name)
 }
 
+/// Handles ensuring the name of rewritten expressions is not changed.
```
Comment (alamb) [inline datafusion/expr/src/expr_rewriter/mod.rs:282]: FYI @findepi here is an example of using Expr::Alias as described in https://github.com/apache/datafusion/issues/1468#issuecomment-2308237517

**Review by alamb — APPROVED** [review]
👏  -- this is a very elegant solution @jonahgao. Thank you very much.

FYI @karlovnv

**Comment by karlovnv** [comment]
@jonahgao @alamb Amazing!
A query like ```select count (case when f_1 > $1 then f_0 else null end) from table1``` for now is working as well!

**Comment by alamb** [comment]
Thanks again @jonahgao

**Pull request merged**